dvc/objects/__init__.py

```diff
@@ -38,18 +42,26 @@ def save(
             ):
                 future.result()
 
+            # if dir cache for this tree has already been staged, move the
+            # staged object rather than uploading/copying a new object
+            staging = get_staging(odb)
+            if staging.exists(obj.hash_info):
+                obj = staging.get(obj.hash_info)
+                kwargs["move"] = True
```

Comment:
This could also be a flag for `save()` instead of the exists check, indicating that we already know the object has been explicitly staged (since `obj = objects.stage()` `save(odb, obj)` is a common pattern)